### PR TITLE
[IS 5.2.0] OIDC logout and Endpoints

### PR DESCRIPTION
**Pull request by DMHP**

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/oauth2_logout.jsp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            </div>
+        </div>
+    </div>
+</div>
+<div class="container main-login-container" style="margin-top:10px;">
+    <div class="row">
+        <div class="span12 content-section">
+            <h3 style="text-align:left;margin-bottom:10px;">Do You really want to logout of Identity Server?</h3>
+            <button id="ok" class="btn btn-primary btn-large" onclick="ok()">OK</button>
+            <button id="cancel" class="btn btn-primary btn-large" onclick="cancel()">Cancel</button>
```
Comment (thariyarox) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/oauth2_logout.jsp]: when using button tag, different browsers treat different default button types. So it is better to define the button type [1]. 

[1] http://www.w3schools.com/tags/tag_button.asp

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java [hunk 2]
```diff
@@ -165,4 +165,17 @@ private OAuthError(){
     private OAuthConstants(){
 
     }
+    public static final String IDP_NAME="LOCAL";
+    public static final String RP_ORIGIN="http://localhost:8080";
```
Comment (thariyarox) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Why RP_ORIGIN is a hard coded url ? It's better if this is configurable.

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 3]
```diff
@@ -60,14 +60,8 @@
...
 import javax.servlet.http.HttpServletResponse;
-import javax.ws.rs.Consumes;
-import javax.ws.rs.GET;
-import javax.ws.rs.POST;
-import javax.ws.rs.Path;
-import javax.ws.rs.Produces;
-import javax.ws.rs.core.Context;
-import javax.ws.rs.core.MultivaluedMap;
-import javax.ws.rs.core.Response;
+import javax.ws.rs.*;
```
Comment (thariyarox) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: It's not the best practice to import *. instead import one by one.

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -547,6 +566,8 @@ private String handleOAuthAuthorizationRequest(String clientId, HttpServletReque
             sessionDataCacheEntryNew.setParamMap(new ConcurrentHashMap<String, String[]>(req.getParameterMap()));
         }
         SessionDataCache.getInstance().addToCache(cacheKey, sessionDataCacheEntryNew);
+        LogoutEndpoint logoutEndpoint=new LogoutEndpoint();
```
Comment (thariyarox) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Code is not formatted. Use the WSO2 code style template and format.

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/LogoutEndpoint.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ */
+package org.wso2.carbon.identity.oauth.endpoint.session;
+
+import org.wso2.carbon.identity.oauth.common.OAuth2ErrorCodes;
+import org.wso2.carbon.identity.oauth.common.OAuthConstants;
+import org.wso2.carbon.identity.oauth.endpoint.util.EndpointUtil;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.*;
```
Comment (thariyarox) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/LogoutEndpoint.java]: Do not import *. Import only the required ones one by one.

**Comment by hpmtissera** [comment]
5.2.0 Feature

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 6]
```diff
@@ -182,8 +190,19 @@ public Response authorize(@Context HttpServletRequest request) throws URISyntaxE
                         sessionDataCacheEntry.setAuthenticatedIdPs(authnResult.getAuthenticatedIdPs());
                         SessionDataCache.getInstance().addToCache(cacheKey, sessionDataCacheEntry);
                         redirectURL = doUserAuthz(request, sessionDataKeyFromLogin, sessionDataCacheEntry);
-                        return Response.status(HttpServletResponse.SC_FOUND).location(new URI(redirectURL)).build();
-
+                        if (redirectURL.contains("oauth2_consent")) {
```
Comment (hpmtissera) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: define constants

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java [hunk 7]
```diff
@@ -165,4 +165,14 @@ private OAuthError(){
     private OAuthConstants(){
 
     }
+
+    public static final String ID_TOKEN_HINT = "&id_token_hint=";
+    public static final int STATUS_OK = 200;
+    public static final String MSG = "Invalid logout request";
+    public static final String RP_STATUS_LOGOUT = "True";
```
Comment (isharak) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Variable name and its value does not give any sense, better to use meaning full names

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java [hunk 8]
```diff
@@ -165,4 +165,14 @@ private OAuthError(){
...
+    public static final String RP_STATUS_LOGGED = "False";
+    public static final String EMPTY_STRING = "";
+    public static final String LOGOUT_HTML = "<html><head><script>window.open(\"*\")</script></head><body></body></html>";
+    public static final String STATUSCOOKIE = "statusCookie";
+    public static final String LOG_OUT_NOTIFICATION = "T";
```
Comment (isharak) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: What is the use of this LOG_OUT_NOTIFICATION

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java [hunk 9]
```diff
@@ -165,4 +165,14 @@ private OAuthError(){
     private OAuthConstants(){
 
     }
+
+    public static final String ID_TOKEN_HINT = "&id_token_hint=";
+    public static final int STATUS_OK = 200;
+    public static final String MSG = "Invalid logout request";
+    public static final String RP_STATUS_LOGOUT = "True";
+    public static final String RP_STATUS_LOGGED = "False";
+    public static final String EMPTY_STRING = "";
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: No need to define a constant for this, can use the StringUtils.EMPTY

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    private static String callBackurl = null;
+    private static String id_token_hint = null;
+
+    @GET
+    @Path("/")
+    @Produces("text/html")
+    public String getLoginStatus(@Context HttpServletRequest request, @Context HttpServletResponse resp, @QueryParam("logoutNotification") String logoutNotification) {
+        String callBackUrl;
+        callBackUrl = getCallBackUrl();
+        if (logoutNotification.equals(OAuthConstants.LOG_OUT_NOTIFICATION)) {
```
Comment (GayanM) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java]: check equals, constant first

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+    /**
+     * @param request which is the request of the endpoint
+     * @return value of the statusCookie
+     */
+    private static String getCookie(HttpServletRequest request) {
+        Cookie[] cookies = request.getCookies();
+        if (cookies != null) {
+            for (javax.servlet.http.Cookie cookie : cookies) {
+                if (cookie.getName().equals(OAuthConstants.STATUSCOOKIE)) {
```
Comment (GayanM) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java]: check equals, constant first

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            }
+        }
+        return null;
+    }
+
+    public static void removeStatusCookie(HttpServletRequest req, HttpServletResponse resp) {
+        Cookie[] cookies = req.getCookies();
+        if (cookies != null) {
+            for (Cookie cookie : cookies) {
+                if (cookie.getName().equals(OAuthConstants.STATUSCOOKIE)) {
```
Comment (GayanM) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java]: check equals, constant first

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/oauth2_logout.jsp (hunk 1 above)
Comment (DMHP) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/oauth2_logout.jsp]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 4 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 3 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java (hunk 2 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 6 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java (hunk 7 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java (hunk 8 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java (hunk 9 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java (hunk 10 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java (hunk 11 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java (hunk 12 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java (hunk 11 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java (hunk 10 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/session/OIDCSessionMgtEndpoint.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java (hunk 9 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java (hunk 8 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java (hunk 7 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: Done

File: components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 6 above)
Comment (DMHP) [inline components/oauth/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Done

**Pull request closed without merging**